zabbixでログから数値を抽出して監視する (#91)
mseninc/blog: norikazum/2022/extract-numerical-values-from-logs-with-zabbix-and-monitor/index.md

---
title: Zabbixでログから数値を抽出して監視する
date: 
author: norikazum

> tags: [zabbix]

Reviewer: ```suggestion
tags: [Zabbix]
```

description: 
---

こんにちは。

今回は、Zabbixを使用しログから数値を抽出して監視する方法を紹介します。

[過去の記事](https://mseeeen.msen.jp/install-clam-antivirus-on-amazon-linux-2-to-automate-checks/)で、ClamAVを利用してログファイルを出力する方法を紹介しました。これを応用しマルウェアが検知されたことをZabbixで検知します。

## 環境
- Zabbix server 6.0.1

![](images/2022-04-09_23h42_46.jpg)

ログファイルが Zabbix ユーザーからアクセスできる必要があります。
環境によっては、権限の調整が必要になりますので以下の記事あたりを参考に設定してください。

[技術メモメモ: ZabbixでAllowRoot=1をせずに/var/log/messagesなどを監視する方法](https://tech-mmmm.blogspot.com/2018/03/zabbixallowroot1varlogmessages.html)

## 監視仕様
- `scanvirus.log` から `Infected files: 数字` を検出する。

> - 数字の部分を抜き出して、`0 を超える` のとき通報する。

Reviewer: ```suggestion
- 数字の部分を抜き出して、`0 を超える` とき通報する。
```

※`scanvirus.log` は [過去の記事](https://mseeeen.msen.jp/install-clam-antivirus-on-amazon-linux-2-to-automate-checks/) を `clamav_check_YYYYMMDD-HHMMSS.log` のファイル名を変更して応用しています。

## Zabbix の設定
それでは、Zabbixの設定をしていきましょう。

流用できるよう、テンプレートを作成します。
`設定→テンプレート→テンプレートの作成` をクリックします。

`テンプレート名`、`グループ` を任意に入力し`追加`します。
記事中は、`Scan Virus Log Infected files Check` とします。

### アイテム
以下のアイテムを作成します。

- 名前：マルウェアファイル検出監視
- タイプ：Zabbixエージェント(アクティブ)

> - キー：log[/var/log/scanvirus.log,"Infected files: ([0-9]+)",,,skip,\1]

Reviewer: ```suggestion
- キー：`log[/var/log/scanvirus.log,"Infected files: ([0-9]+)",,,skip,\1]`
```

- データ型：ログ
- 監視間隔：1m

![マルウェアファイル検出監視用アイテムの作成](images/2022-04-09_23h54_44.jpg)

log 関数のフォーマットは、`log[/path/to/file/file_name,<regexp>,<encoding>,<maxlines>,<mode>,<output>]` となっています。

第2引数の、`regexp` はログの取得対象の文字列を記載します。
本記事では、マルウェアファイルの検出ファイル数を取り出したいので `([0-9]+)` と `1桁以上の連続する数字` を示す正規表現を記載します。

第6引数の、`output` は `\1` を指定します。
これは **第2引数の()で囲んだ箇所をアイテムの値として抽出する** という設定になります。

## トリガー
以下のトリガーを作成します。

- 名前：Found infected files : {HOST.NAME}
- 深刻度：重度の障害
- 条件式：`last(/Scan Virus Log Infected files Check/log[/var/log/scanvirus.log,"Infected files: ([0-9]+)",,,skip,\1])>=1`

![](images/2022-04-09_23h59_52.jpg)

> 条件式は前述で作成したアイテムを指定していますが、最後の `>=1` が `1以上` を表します。

Reviewer: ```suggestion
条件式は前述で作成したアイテムを指定しています。最後の `>=1` で `1以上` を表します。
```

論理的には意味は同じなんですが、冒頭には「0を超える」という表現で書いてあったのに、ここでは「1以上」となっているので前者に統一したほうがいいんじゃないでしょうか。

Reviewer: ↑どうでしょう

Author: 遅くなりました。0 を超える　に統一しました。

---
title: zabbixでログから数値を抽出して監視する
date: 
author: norikazum

> tags: [zabbix]

Reviewer: 他の記事って Zabbix って頭大文字じゃなかったですか？

description: 
---

こんにちは。

> 今回は、Zabbixを資料しログから数値を抽出して監視する方法を紹介します。

Reviewer 2: > 資料し

使用し？

> [過去の記事](https://mseeeen.msen.jp/install-clam-antivirus-on-amazon-linux-2-to-automate-checks/)で、ClamAVを利用してログファイルを出力する方法を紹介しましたが、これを応用しマルウェアが検知されたことをzabbixで検知したいと思います。

Reviewer: ```suggestion
[過去の記事](https://mseeeen.msen.jp/install-clam-antivirus-on-amazon-linux-2-to-automate-checks/)で、ClamAVを利用してログファイルを出力する方法を紹介しました。これを応用しマルウェアが検知されたことをzabbixで検知します。
```

「が」が使われていますが、逆接になっていません。また、「～思います」は自分の感想など以外では、別の表現に変えてみてください。

## 環境
- Zabbix server 6.0.1

![](images/2022-04-09_23h42_46.jpg)

ログファイルが zabbix ユーザーからアクセスできる必要があります。
環境によっては、権限の調整が必要になりますので以下の記事あたりを参考に設定してください。

> [技術メモメモ: ZabbixでAllowRoot=1をせずに/var/log/messagesなどを監視する方法](https://tech-mmmm.blogspot.com/2018/03/zabbixallowroot1varlogmessages.html)

Reviewer: 文章にくっついて見えるので、空行を入れるか、箇条書きにしてもらえますか

```suggestion

- [技術メモメモ: ZabbixでAllowRoot=1をせずに/var/log/messagesなどを監視する方法](https://tech-mmmm.blogspot.com/2018/03/zabbixallowroot1varlogmessages.html)
```

## 監視仕様

> - `scanvirus.log` から `Infected files: 数字` を検出する。

Reviewer 2: > `scanvirus.log` から `Infected files: 数字` を検出する。

このログが何なのか説明もう少し欲しいです。
過去記事にも記載ありませんでした。

> - 数字の部分を抜き出して、`0以外` のとき通報する。

Reviewer 2: 後述していますが、マイナスもあり得るので 0 を超える場合の方がいいかと思います。

```suggestion
- 数字の部分を抜き出して、`0 を超える` のとき通報する。
```

もしくは

```suggestion
- 数字の部分を抜き出して、`1 以上` のとき通報する。
```

Reviewer 2: すみません、間違いありました。下記ですね。

```suggestion
- 数字の部分を抜き出して、`0 を超える` とき通報する。
```

## zabbix の設定
それでは、zabbixの設定をしていきましょう。

流用できるよう、テンプレートを作成します。
`設定→テンプレート→テンプレートの作成` をクリックします。

`テンプレート名`、`グループ` を任意に入力し`追加`します。
記事中は、`Scan Virus Log Infected files Check` とします。

### アイテム
以下のアイテムを作成します。

```
名前：マルウェアファイル検出監視
タイプ：Zabbixエージェント(アクティブ)
キー：log[/var/log/scanvirus.log,"Infected files: ([0-9]+)",,,skip,\1]
データ型：ログ
監視間隔：1m

> ```

Reviewer: これはコード的なものではないので、箇条書きかテーブルに変更してください。

```suggestion
- 名前：マルウェアファイル検出監視
- タイプ：Zabbixエージェント(アクティブ)
- キー：`log[/var/log/scanvirus.log,"Infected files: ([0-9]+)",,,skip,\1]`
- データ型：ログ
- 監視間隔：1m
```

> ![](images/2022-04-09_23h54_44.jpg)

Reviewer: 画像には説明をいれてください

```suggestion
![マルウェアファイル検出監視用アイテムの作成](images/2022-04-09_23h54_44.jpg)
```

log 関数のフォーマットは、`log[/path/to/file/file_name,<regexp>,<encoding>,<maxlines>,<mode>,<output>]` となっています。

第2引数の、`regexp` はログの取得対象の文字列を記載します。
本記事では、マルウェアファイルの検出ファイル数を取り出したいので `([0-9]+)` と `1桁以上の連続する数字` を示す正規表現を記載します。

第6引数の、`output` は `\1` を指定します。
これは **第2引数の()で囲んだ箇所をアイテムの値として抽出する** という設定になります。

## トリガー
以下のトリガーを作成します。

```
名前：Found infected files : {HOST.NAME}
深刻度：重度の障害
条件式：`last(/Scan Virus Log Infected files Check/log[/var/log/scanvirus.log,"Infected files: ([0-9]+)",,,skip,\1])>=1`

> ```

Reviewer: 同様に箇条書きのほうがいいと思います。

![](images/2022-04-09_23h59_52.jpg)

> 条件式はアイテムを指定していますが、最後の `>=1` が 数字の部分が `0以外` となります。

Reviewer 2: > 条件式はアイテムを指定していますが、最後の `>=1` が 数字の部分が `0以外` となります。

こちら、この条件式だとマイナストリガーされないので1以上だと思います。

Reviewer: > 条件式はアイテムを指定していますが、

これってどういう意味ですか？

> 最後の >=1 が 数字の部分が 0以外 となります。

ちょっとこれも意味がわかりづらいのですが、「最後の >=1 が 0以外 を表します。」ということですか？

## あとがき
本記事の設定で、マルウェアが1つ以上検出された場合に検出されます。

> 今回の記事は自動復旧ではなく、手動クローズを想定しています。

Reviewer 2: > 

条件式を見る限り、次回のチェック（最新の値）で Infected files が 0 だった場合、復旧されそうなのですが検証済みでしょうか？

Reviewer 3: 検証していません。
後書きに、「今回の記事は自動復旧ではなく、手動クローズを想定しています。」としています。
